client/my-sites/plans/current-plan/index.jsx

```diff
@@ -191,12 +191,19 @@ class CurrentPlan extends Component {
 
 		if (
 			eligibleForProPlan &&
-			! isFlexiblePlanProduct( selectedSite.plan ) &&
+			! isFreePlanProduct( selectedSite.plan ) &&
 			! isPro( selectedSite.plan )
 		) {
 			showLegacyPlanNotice = true;
 		}
 
+		// Ensures the Plan tab is shown in case the plan changes after the controller redirect.
+		if ( eligibleForProPlan && isFreePlanProduct( selectedSite.plan ) ) {
+			page.redirect( `/plans/${ selectedSite.slug }` );
+
+			return null;
+		}
+
```

Comment:
It works. However, there is a very similar redirect logic in the controller:
https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans/current-plan/controller.jsx#L23

From my test, changing:
```
if (
		isFreePlanProduct( selectedSite.plan ) ||
		( eligibleForProPlan && isFlexiblePlanProduct( selectedSite.plan ) )
	)
```
into what you wrote here:
`if ( eligibleForProPlan && isFreePlanProduct( selectedSite.plan ) )`
works similarly as well, i.e. there is also a flash of the Free plan page due to the nature of redirecting. 

Since controller is where most of the routing logics happen, what do you think if we move this over there?

Reply:
@southp  Thanks for reviewing and bringing this up!

> if ( eligibleForProPlan && isFreePlanProduct( selectedSite.plan ) )

I'm afraid changing to this condition in the controller would discontinue the redirect that was happening on the Legacy Plans. On the other hand, we could just keep `isFreePlanProduct( selectedSite.plan )` and it should cover the overhauled Plans as well.

> Since controller is where most of the routing logics happen, what do you think if we move this over there?

This was actually the root of the problem: relying on the controller only works when Plan doesn't change right before visiting the Plans and My Plan pages. It appears that the Pro Plan is still active after cancellation for the controller files. This was true before with the Legacy, but it became visible only after adding the additional redirect from Plans to My Plan in #61656.

So what happening now:
1. Pro plan is cancelled
2. You click on Plans button
3. Landing on /plans, the controller still sees Pro plan (selectedSite.plan)
4. Redirects to /my-plan (added in #61656)
5. My Plan controller also sees the Pro plan and is fine with it - no redirect.
6. This is where we need the extra redirect on index where the updated site info is fetched and the component, eventually, realises the plan has changed and then redirects back to /plans.

Alternatively, we could just add the redirect on the /plans' index page in the first place and avoid the back-and-forth, but this would make the redirect unnecessarily slow for most of the other cases where the plan doesn't change.